crates/volta-core/src/tool/package/fetch.rs

```diff
@@ -0,0 +1,143 @@
+//! Provides fetcher for 3rd-party packages
+
+use std::fs::{rename, write, File};
+use std::io::{Read, Seek, SeekFrom};
+use std::path::{Path, PathBuf};
+
+use super::super::download_tool_error;
+use crate::error::ErrorDetails;
+use crate::fs::{
+    create_staging_dir, ensure_containing_dir_exists, ensure_dir_does_not_exist, read_dir_eager,
+    read_file,
+};
+use crate::path;
+use crate::style::{progress_bar, tool_version};
+use crate::tool::{self, PackageDetails};
+use crate::version::VersionSpec;
+use archive::{Archive, Tarball};
+use log::debug;
+use semver::Version;
+use sha1::{Digest, Sha1};
+use volta_fail::{Fallible, ResultExt};
+
+pub fn fetch(name: &str, details: &PackageDetails) -> Fallible<()> {
+    let version_string = details.version.to_string();
+    let cache_file = path::package_distro_file(&name, &version_string)?;
+    let shasum_file = path::package_distro_shasum(&name, &version_string)?;
+
+    if let Some(archive) = load_cached_distro(&cache_file, &shasum_file) {
+        debug!(
+            "Loading {} from cached archive at '{}'",
+            tool_version(&name, &version_string),
+            cache_file.display(),
+        );
+        unpack_archive(archive, name, &details.version)
+    } else {
+        let archive = fetch_remote_distro(
+            tool::Spec::Package(name.into(), VersionSpec::exact(&details.version)),
+            &details.tarball_url,
+            &cache_file,
+        )?;
+
+        unpack_archive(archive, &name, &details.version)?;
+
+        // Save the shasum in a file
+        write(&shasum_file, details.shasum.as_bytes()).with_context(|_| {
+            ErrorDetails::WritePackageShasumError {
+                package: name.into(),
+                version: version_string,
+                file: shasum_file,
+            }
+        })
+    }
+    // Check for cache, if available then use that (unpack_archive)
```

Comment:
Looks like this comment should probably move to the top of the function and/or be deleted wholesale?

Reply:
Whoops, that should have been deleted! That was one of my comments I did mapping out the flow of the code before I started writing 😓Good catch!